Show online players on servers screen

iTrooz opened the pull request:
<!--
Hey there! Thanks for your contribution.

Please make sure that your commits are signed off first.
If you don't know how that works, check out our contribution guidelines: https://github.com/PrismLauncher/PrismLauncher/blob/develop/CONTRIBUTING.md#signing-your-work
If you already created your commits, you can run `git rebase --signoff develop` to retroactively sign-off all your commits and `git push --force` to override what you have pushed already.

Note that signing and signing-off are two different things!
-->

iTrooz committed:
Add dummy UI button

iTrooz committed:
add my classes

iTrooz committed:
add method queryStatus() to servers and use it on click

iTrooz committed:
add my classes to CMakeLists

iTrooz committed:
remove space

iTrooz committed:
add debug print

iTrooz committed:
make splitAddress() const

iTrooz committed:
run the code in tasks

iTrooz committed:
make MCResolver a dynamic object so it doesnt get deleted before the callback

See https://discord.com/channels/1031648380885147709/1031823065937629267/1307471566166167696

iTrooz committed:
replace my printf calls with qDebug

iTrooz committed:
Actually show online players when clicking on the button

iTrooz committed:
use std::optional<> to signify when there is no value

iTrooz committed:
No need to close the socket, it is done automatically

https://doc.qt.io/qt-6/qtcpsocket.html#dtor.QTcpSocket

iTrooz committed:
replace std::runtime_exception with PrismLauncher Exception + add try/catch

iTrooz committed:
fix warning about unused jsonLength

iTrooz committed:
avoid translating back to stf string without reason

iTrooz committed:
Fix after rebase

iTrooz committed:
fix qDebug() call with string << append

iTrooz committed:
use Qt5-compatible writeString() impl

iTrooz committed:
refactor header and code into 2 files

iTrooz committed:
MCResolver -> McResolver

iTrooz committed:
.hpp -> .h

iTrooz committed:
remove useless code when reading response

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -112,11 +125,42 @@ struct Server {
+class ServerPingTask : public Task {
+    Q_OBJECT
+  public:
+    explicit ServerPingTask(QObject* parent, Server &server) : Task(parent), m_server(server) {}
```
```suggestion
    explicit ServerPingTask(Server &server) : Task(), m_server(server) {}
```

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -112,11 +125,42 @@ struct Server {
+    }
+    
+  private:
+    Server &m_server;
```
This may cause a crash

iTrooz replied:
Could you explain your reasoning ? Using `m_server` in the `QObject::connect()` callback is indeed causing a crash right now (only sometimes, so I assume it's a race condition), but I cannot figure out why that would happen

Trial97 replied:
you pass a reference that reference is valid as long as the original reference is valid.
So once the server object that you passed to task is destructed this variable will not point to a correct memory address.
So what can cause the crash is as follows:
- the server is created on the ServerPage(and maybe associated with it)
- you pass the reference to this task(right now a pointer)
- task is executed async
- the serverPage is destroyed and together with it the server object
- but because you used a task pointer there are a few cases where they remain until the task is finished( e.g. you passed to the lambda a reference to that pointer to deleteLater)
- the task still has an invalid reference

PS. This is a speculation and may not be valid for you case but this is a recent crash caused by the capture of the tasks as a shared pointer.

iTrooz replied:
This is not what's happening in this specific case since I don't leave the server tab, but that's still something I'll have to take into consideration, thanks !

iTrooz replied:
Ok, found out what's happening: the `ServerPingTask` object is being deleted before the task finishes, probably because `Task::Ptr()` immediately calls `deleteLater()` on the object without waiting for it to finish.

I'll try to figure out a solution to this, probably just by linking the finished() signal of the task to deleteLater()

iTrooz replied:
In the end, the reason was that `job` was deleted a the end of the function call (what you said in https://github.com/PrismLauncher/PrismLauncher/pull/3112#discussion_r1846914296), causing the tasks to be destroyed, even though the they were not completed

Trial97 replied:
I still want this to be something else not a reference: is a bad idea to modify a reference async and there is no way to check if is still valid

iTrooz replied:
Done ! With 7c61fec8e0e753de1b7215e45b0a4272551ffac4 and fef8ee2d1b4f866437b30ec4ea8fd5cae2426e9c
Just tried, and closing the popup indeed stops `server` from being modified once the task has finished

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -433,6 +481,16 @@ class ServersModel : public QAbstractListModel {
+            ServerPingTask *task = new ServerPingTask(this, server);
+            job->addTask(Task::Ptr(task));
+        }
+        job->start();
```
This looks like if  is not blocking the task will not be executed as `job` will be destructed

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -112,11 +125,42 @@ struct Server {
+    virtual void executeTask() override {
+        qDebug() << "Querying status of " << m_server.m_address;
+        auto [domain, port] = m_server.splitAddress();
+        McResolver *resolver = new McResolver(nullptr, domain, port);
```
This is a leak: either use smart pointers as class members or connect the fail signal to deleteLater

iTrooz replied:
Shouldn't adding a McResolver instance as a class member (without using a smart pointer) suffice ?

~~Btw, could you explain to me what a Task is, and if I should even use one here ? Right now I don't see its utility at all, I could call my code directly~~

Trial97 replied:
From my knowledge QObject can't be  moved so if you do not use pointers/smartpointers then you need to initialize it when you construct the object

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -112,11 +125,42 @@ struct Server {
+        qDebug() << "Querying status of " << m_server.m_address;
+        auto [domain, port] = m_server.splitAddress();
+        McResolver *resolver = new McResolver(nullptr, domain, port);
+        QObject::connect(resolver, &McResolver::succeed, [=](QString ip, int port) {
```
please do not use `[=]` (capture all) in lambda

iTrooz replied:
Sure !

Could you explain to me why this shouldn't be used ? I thought `[=]` only mean "capture implicitly what is used", not "capture all"

Ryex replied:
it leads to some situation where your using something implicitly you didn't intend to. like the `this` pointer, or a shared pointer to a qobject the signals bound to causing circular references.
it's just best practice with async lambdas to be explicit as possible about what's getting captured.

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -734,4 +792,10 @@ void ServersPage::on_actionJoin_triggered()
+void ServersPage::on_actionRefresh_triggered()
+{
+    qDebug() << "Action clicked";
+    m_model->queryServersStatus();
```
It would be nice not to allow multiple calls.
Also consider doing this on page start also

iTrooz replied:
> Also consider doing this on page start also

Done

> It would be nice not to allow multiple calls.

If someone has a shut down server in their list, that would disable the "refresh" button until the network timeout, which may be a bad thing. In the game multiplayer screen, you can spam the "refresh" button to refresh servers even if it hasn't finished to ping every server.
So, how about keeping the current behaviour ?

Trial97 replied:
Then I would like to stop somehow the previous call:
if the user clicks multiple times the previous task should be aborted and the signals disconnected.
That way you get only the latest signals

Trial97 replied:
Please consider a mechanism to delete/stop the task that started. In case of multiple hits on refresh the task will accumulate in memory

iTrooz replied:
Done ! ae7d3379e4ed2d8aee51026f37b7344026f97de7
 
Tested using `sudo tc qdisc add dev wlan0 root netem delay 5000ms` to manually slow down network, the old task's destructor is called instantly when refreshing

iTrooz committed:
remove QObject parent argument from ServerPingTask constructor

iTrooz committed:
remove debug print

iTrooz committed:
ping servers when opening Servers page

iTrooz committed:
do not capture things implicitely in lambdas

iTrooz committed:
fix `job` being deleted before tasks are finished

iTrooz committed:
communicate when ServerPingTask succeeds/fails

iTrooz committed:
always delete McResolver object, even when it fails

iTrooz committed:
cleanup

iTrooz committed:
make McResolver have the same signal name conventions as Task

iTrooz committed:
add documentation for Task and ConcurrentTask

iTrooz committed:
run socket code in thread

iTrooz committed:
put more McClient methods to private

iTrooz committed:
refresh UI when we got the players online

Trial97 commented on launcher/tasks/Task.h:
```diff
@@ -161,11 +170,17 @@ class Task : public QObject, public QRunnable {
    protected slots:
+    //! The Task subclass must call this method when the task has succeeded 
     virtual void emitSucceeded();
     virtual void emitAborted();
```
//! The Task subclass must call this method when the task was aborted

iTrooz replied:
The thing I wasn't sure about: external code can also call this method, right ? Any piece of code can ask a task to abort ?

Also, what's the difference between `emitAborted()` and `abort()` ? When should one be used over the other ?

Trial97 replied:
abort also checks the canAbort() function and should always be called in external code.
this function should be called by the internal task code

Trial97 commented on launcher/ui/pages/instance/McClient.cpp:
```diff
@@ -0,0 +1,165 @@
+    return QtConcurrent::run([this]() {
+        try {
+            auto status = getStatusDataBlocking();
+            int onlinePlayers = status.value("players").toObject().value("online").toInt();
```
You can adapt this to use the https://github.com/PrismLauncher/PrismLauncher/blob/develop/launcher/Json.h functions like:`requireInt`

iTrooz replied:
What is the advantage of this, other than being quicker to write ?

Trial97 commented on launcher/ui/pages/instance/McClient.cpp:
```diff
@@ -0,0 +1,165 @@
+
+// From https://wiki.vg/Protocol#VarInt_and_VarLong
+void McClient::writeVarInt(QByteArray &data, int value) {
+    while (true) {
```
can you replace the infinite loop?

Trial97 commented on launcher/ui/pages/instance/McClient.cpp:
```diff
@@ -0,0 +1,165 @@
+    int position = 0;
+    char currentByte;
+
+    while (true) {
```
can you replace the infinite loop?

Trial97 commented on launcher/ui/pages/instance/McResolver.cpp:
```diff
@@ -0,0 +1,74 @@
+    lookup->setName(QString("_minecraft._tcp.%1").arg(domain));
+    lookup->setType(QDnsLookup::SRV);
+
+    connect(lookup, &QDnsLookup::finished, this, [&, domain, port]() {
```
```suggestion
    connect(lookup, &QDnsLookup::finished, this, [ domain, port]() {
```

Trial97 commented on launcher/ui/pages/instance/McResolver.cpp:
```diff
@@ -0,0 +1,74 @@
+}
+
+void McResolver::pingWithDomainA(QString domain, int port) {
+    QHostInfo::lookupHost(domain, this, [&, port](const QHostInfo &hostInfo){
```
```suggestion
    QHostInfo::lookupHost(domain, this, [port](const QHostInfo &hostInfo){
```

Trial97 commented on launcher/ui/pages/instance/McResolver.cpp:
```diff
@@ -0,0 +1,74 @@
+        if (hostInfo.error() != QHostInfo::NoError) {
+            emitFail("A record lookup failed");
+            return;
+        } else {
```
```suggestion
        } 
```

Trial97 commented on launcher/ui/pages/instance/McResolver.cpp:
```diff
@@ -0,0 +1,74 @@
+            
+            const auto& firstRecord = records.at(0);
+            emitSucceed(firstRecord.toString(), port);
+        }
```
```suggestion
```

Trial97 commented on launcher/ui/pages/instance/McResolver.h:
```diff
@@ -0,0 +1,28 @@
+private:
+    void pingWithDomainSRV(QString domain, int port);
+    void pingWithDomainA(QString domain, int port);
+    void emitFail(std::string error);
```
maybe:
```suggestion
    void emitFail(QString error);
```

iTrooz replied:
Oh, yeah.
I'm not used to a library having this much impact over the codebase. At first, I was trying to fully isolate Qt from theses classes, before realising Qt was an integral part of the project, not just a simple library for the GUI that you can isolate from the other parts.

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -112,11 +126,62 @@ struct Server {
     int m_maxPlayers = 0;
 };
 
+class ServerPingTask : public Task {
```
consider moving the task to separate file

iTrooz replied:
Sure ! Where should that file be ? In `launcher/ui/pages/instance`, or somewhere else ?
I'd also love to move McResolver and McClient there, I feel like they don't belong in their current folder

Trial97 replied:
yeah in the `launcher/ui/pages/instance/` it should be on the same level as the other code added e.g. `McClient.cpp`

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -433,6 +502,25 @@ class ServersModel : public QAbstractListModel {
+        }
+        job->start();
+
+        connect(job, &ConcurrentTask::finished, [job]() {
```
Move this connect above start

iTrooz committed:
Make task output online players rather than updating Server itself

This doesn't really change anything

iTrooz committed:
Disconnect task finished signal when ServersModel is destroyed

iTrooz committed:
Do not timeout after just 3000ms

Keep the network default timeout

iTrooz committed:
remove implicit captures

iTrooz committed:
simplify code

iTrooz committed:
replace std::string by QString in emitFail()

iTrooz committed:
connect job to deletion task before starting it

This ensures the signal is not fired before we have connected it

Eragoneq commented:
I was considering to try and implement a feature, where servers could send pack metadata that could then be used to automatically create an instance with the given mods and configuration.

I think you're already fetching the response from the server with the JSON data and only reading the number of players, could it be also reused in the scenario I described?

Also question to the maintainers commenting, could that be a new included feature if I would be able to provide the code for it?

iTrooz commented:
> I think you're already fetching the response from the server with the JSON data and only reading the number of players, could it be also reused in the scenario I described?

Sure, you could modify the code to do that ! You would probably just need to refactor the "get online players" part out of McClient, and do it in the server page instead, and then you'd have access to the whole JSON instead of just the online players, to do whatever you want

iTrooz commented:
This PR fixes https://github.com/PrismLauncher/PrismLauncher/issues/271 btw

iTrooz committed:
remove infinite loop in writeVarInt()

iTrooz committed:
remove infinite loop from readVarInt()

iTrooz committed:
move ServerPingTask in its own file

iTrooz committed:
add documentation about task abortion

iTrooz committed:
cleanup

iTrooz committed:
use Json::requireInteger()

iTrooz committed:
Always autodelete signal connections with `this` captured when the objet is deleted, just to be sure

iTrooz committed:
fix typo

iTrooz committed:
abort current query task if present when refreshing

iTrooz committed:
chore

iTrooz committed:
make writePacketToSocket() clear data automatically

iTrooz committed:
Reset online players UI component when refreshing

iTrooz commented:
Ok, all conversations have been resolved, and the feature feels complete to me. I think this PR can be considered for merging

iTrooz committed:
DCO Remediation Commit for iTrooz <hey@itrooz.fr>

I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: fac521a312d1f6055b29d12da8b8773adaa24afb
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 43a54cafef95a3b4c2181f4d3d1e2d3876b8e7f9
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: ee35ac5afdb2a6409b7ad52fde38c411d112f064
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 99ac11bc408b5102a7c9d6d0c003e854d5e7fdef
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 8fa1dff17dc4d7953fbf7f0961a33e751183b06b
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 2f70115be5555c7c9ad46765f61e6b1cc965e8b4
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: ea2a2349f8ebd353ac4ba4296f9514c3f8419c58
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 87c9066a2b218aef5296ab8ad4dddfa483dcd9cb
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: fe28a051d533bd34681c6306deab5b93b13dc5e9
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 0a379a05ff03b2c55fcc4342c84f90fff64bb9a9
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 6a7678a6e9548501b84ed3a2d81a9cabe6955703
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: cba7e2dc362238cbfa9bcc680d164c828cd39836
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 8cf0c2029ccccf40cbf21fd122bea233732911bc
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 0d830e56e9c888b860e03ee4a9e44239d67a4d4a
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: b35cffb3477b997a4dc779696aeb0acb79880d51
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 1f094b98039b4f8c79bc402246babd4f821116db
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 2d06e0a11194a3a4c379ddd518b857c4c2b98f54
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: c3543b104b5f13778677e2244ab93faf6850cb99
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 8b7040d416b46d8011b97b71505778026c2c5ece
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: b8035ca0783ba42efbc7c11cf329c6f4372abbc5
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 9d5727e36b22d2a2eeb17f6faf0bf5b2c01421fd
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 7cf24586c2a9ff8e5dd33a390f83b7f098266881
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 7d04f0ee758cb3c0c9c4273de67424269d63507c
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 60fb922ba2ed56703e5d64626f015e57d5cb0de5
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: a79a66c177194f1761f353e00bc53f1dc653eea3
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 1fb0fe0171fa2db09b296a62d636bab1ea026009
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 5cfb5a6f0db81e850ed3299759ffc131d049f289
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 9ce5eaaa0c8e8ca74dce42e2d287e865905efdf4
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 0c6f78dee2c16820624e2f55e12d903392baad75
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 7c8d2c9b55526db21a9656a65adbe4b2cfabd892
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 24b9815763ce1f81b30c4d422ee0d2dc5660dede
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 7d2da194184180846362af7c5d8e3726db190b88
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: ca6d66970ea52ae95c064babf42f74a1d953717b
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 520d6b0b42eae5b7fb5757cc07d96b53d2f6e7e8
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 4fad298d675102df39fadd0837bc36450068a7dc
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 087ab70143c69de8e8b9c5006393137f42b41f1c
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 7c61fec8e0e753de1b7215e45b0a4272551ffac4
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: fef8ee2d1b4f866437b30ec4ea8fd5cae2426e9c
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 3a9c03098267fd64a27b9bfe62dddccbdffc6cf0
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: f05548f3a36c0badcedec86d6fc5dd7a973bd36e
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 66f36195d8d2d4c362e94dc56239feb42e31fd36
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: ca52d00b8040e27860f931bdf5a6e7ce92c08413
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 26f50f9b818e68fc21c8ae06fc313bfc2e0da6e9
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 873232ebe3b3e41a97f908018dd5ff7fd14be10e
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 8b90a9f2b3e3d58b9725eda081cf209781c13aa6
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: dbb88ca7df6594f65cba5c5e5bd821a0c2605c00
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 6f9be258dc87752c3d8d80cac44570575ecba7b0
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: d124e2e0cba25e914bb885a4efd785294790227c
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: cfb0c97262f2f6467f1f35460938a9160cb2484e
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 5eb417ff2be159330558999082ecb64e3ee8f003
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 3fb6764ea48e942a655aafbbce9fc4a226839eff
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: ae7d3379e4ed2d8aee51026f37b7344026f97de7
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 09782745c6f1e1ad0edfd4222f432d679d10c17b
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: b0778e7a1fcf62493638aa9caafcb17a8860a3a1
I, iTrooz <hey@itrooz.fr>, hereby add my Signed-off-by to this commit: 01db826ec92bd41164514518d25b1941b58e5180

Signed-off-by: iTrooz <hey@itrooz.fr>

Trial97 commented on launcher/tasks/Task.h:
```diff
@@ -161,11 +172,16 @@ class Task : public QObject, public QRunnable {
    protected slots:
+    //! The Task subclass must call this method when the task has succeeded 
     virtual void emitSucceeded();
+    //! **The Task subclass** must call this method when the task has succeeded. External code should call abort() instead. 
```
```suggestion
    //! **The Task subclass** must call this method when the task has aborted. External code should call abort() instead. 
```

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -51,8 +52,10 @@
 #include <QFileSystemWatcher>
 #include <QMenu>
 #include <QTimer>
+#include <QFutureWatcher>
```
```suggestion
```

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -433,6 +448,48 @@ class ServersModel : public QAbstractListModel {
+            currentQueryTask->addTask(Task::Ptr(task));
+
+            // Update the model when the task is done
+            connect(task, &Task::finished, this, [this, task, row, &server]() {
+                server.m_currentPlayers = task->m_outputOnlinePlayers;
+                emit dataChanged(index(row, 0), index(row, COLUMN_COUNT - 1));
+            });
```
```suggestion
            connect(task, &Task::finished, this, [this, task, row]() {
            if (m_servers.size()<row){return;}
            m_server[row].m_currentPlayers = task->m_outputOnlinePlayers;
                emit dataChanged(index(row, 0), index(row, COLUMN_COUNT - 1));
            });
```

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -520,6 +577,7 @@ class ServersModel : public QAbstractListModel {
     QList<Server> m_servers;
     QFileSystemWatcher* m_watcher = nullptr;
     QTimer m_saveTimer;
+    ConcurrentTask *currentQueryTask = nullptr;
```
```suggestion
    ConcurrentTask::Ptr m_currentQueryTask = nullptr;
```

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -433,6 +448,48 @@ class ServersModel : public QAbstractListModel {
+
+        // make task delete itself when done
+        auto *c = currentQueryTask;
+        connect(currentQueryTask, &ConcurrentTask::finished, [c]() {
+            c->deleteLater();
+        });
+
+        // Also delete it from the model, if the model itself hasn't been deleted
+        connect(currentQueryTask, &ConcurrentTask::finished, this, [this, c]() {
+            if (c == currentQueryTask) currentQueryTask = nullptr;
+        });
```
Use shared pointers and get rid of this

Trial97 commented on launcher/ui/pages/instance/McClient.h:
```diff
@@ -0,0 +1,38 @@
+class McClient : public QObject {
+    Q_OBJECT
+
+    QString domain;
+    QString ip;
+    short port;
+    QTcpSocket socket;
```
```suggestion
    QString m_domain;
    QString m_ip;
    short m_port;
    QTcpSocket m_socket;
```

Trial97 commented on launcher/ui/pages/instance/McResolver.h:
```diff
@@ -0,0 +1,28 @@
+class McResolver : public QObject {
+    Q_OBJECT
+
+    QString constrDomain;
+    int constrPort;
```
```suggestion
    QString m_constrDomain;
    int m_constrPort;
```

iTrooz committed:
chore

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
add `m_` prefix to class members

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
Use shared pointer to store server query task

Signed-off-by: iTrooz <hey@itrooz.fr>

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -520,6 +573,7 @@ class ServersModel : public QAbstractListModel {
     QList<Server> m_servers;
     QFileSystemWatcher* m_watcher = nullptr;
     QTimer m_saveTimer;
+    ConcurrentTask::Ptr currentQueryTask = nullptr;
```
```suggestion
    ConcurrentTask::Ptr m_currentQueryTask = nullptr;
```

Trial97 commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -433,6 +447,45 @@ class ServersModel : public QAbstractListModel {
+
+        // Destroy task when done
+        connect(currentQueryTask.get(), &ConcurrentTask::finished, this, [this]() {
+            currentQueryTask = nullptr;
```
Better not do this and leave this to be destroyed together with the ServerPage

iTrooz committed:
currentQueryTask -> m_currentQueryTask

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
do not destroy the ping ConcurrentTask directly at the end of the task

Signed-off-by: iTrooz <hey@itrooz.fr>

TheKodeToad commented on launcher/ui/pages/instance/McClient.cpp:
```diff
@@ -0,0 +1,166 @@
+}
+
+QFuture<int> McClient::getOnlinePlayers() {
+    return QtConcurrent::run([this]() {
+        try {
+            auto status = getStatusDataBlocking();
+            auto players = Json::requireObject(status, "players");
+            return Json::requireInteger(players, "online");
+        } catch (const Exception &e) {
+            qDebug() << "Error: " << e.what();
+            return -1;
+        }
+    });
```
It feels a bit pointless to use a thread when it is just waiting doing nothing while another thread does the actual work...

Also, this method is not recommended on Windows

iTrooz replied:
hey ! Sorry for the long delay

> It feels a bit pointless to use a thread when it is just waiting doing nothing while another thread does the actual work...

In this case the main thread is not really waiting, since `QtConcurrent::run()` returns a future, that is later "awaited" here: https://github.com/PrismLauncher/PrismLauncher/blob/5eb417ff2be159330558999082ecb64e3ee8f003/launcher/ui/pages/instance/ServerPingTask.cpp#L21
If you want, I can try to handle the socket IO operations from the main GUI thread, but it didn't seem like a good option to me (shouldn't the main thread only do GUI stuff ?)

> Also, this method is not recommended on Windows

I do not have much experience with Windows, could you clarify what is the bad pattern here ?

Trial97 replied:
I'm with iTrooz here: 
- it is bad to do this in the UI thread
- we already use a similar pattern when extracting files

TheKodeToad replied:
Pretty sure all the waitFor methods do is make the thread sleep until the work is done. The rest of the work the thread does is just light parsing

> I do not have much experience with Windows, could you clarify what is the bad pattern here ?

https://doc.qt.io/qt-6/qabstractsocket.html#waitForConnected

> Note: This function may fail randomly on Windows. Consider using the event loop and the [connected](https://doc.qt.io/qt-6/qabstractsocket.html#connected)() signal if your software will run on Windows.

iTrooz replied:
> Note: This function may fail randomly on Windows. Consider using the event loop and the [connected](https://doc.qt.io/qt-6/qabstractsocket.html#connected)() signal if your software will run on Windows.

Alright, I changed it. Now it doesn't use another thread anymore

iTrooz committed:
remove unused defines in McClient.h

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
remove waitForConnected() and waitForReadyRead() and use signals ineatd

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
forward McClient error as Task error

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
forward McResolver error as Task error

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
use m_responseReadState to avoid calling parseResponse() (as a failsafe for malicious/bad server responses)

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
update documentation

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
Show socket error in McClient

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz committed:
use more sensible protocol version

Signed-off-by: iTrooz <hey@itrooz.fr>

TheKodeToad commented on launcher/ui/pages/instance/ServersPage.cpp:
```diff
@@ -88,6 +90,15 @@ struct Server {
         }
     }
 
+    std::tuple<QString, int> splitAddress() const {
+        auto parts = m_address.split(":");
+        if (parts.size() == 1) {
+            return std::make_tuple(parts[0], 25565);
+        } else {
+            return std::make_tuple(parts[0], parts[1].toInt());
+        }
+    }
```
I think this is a (less correct) duplication of the logic here:
https://github.com/PrismLauncher/PrismLauncher/blob/develop/launcher/minecraft/launch/MinecraftTarget.cpp#L21-L65

iTrooz replied:
Done !

iTrooz committed:
MinecraftTarget::parse() to parse IP

Signed-off-by: iTrooz <hey@itrooz.fr>

TheKodeToad reviewed (approved):
thank you

Ryex commented:
I'd prefer @getchoo  look at this one too before we merge.

iTrooz committed:
fix typo

Signed-off-by: iTrooz <hey@itrooz.fr>

iTrooz commented:
Any news ?

TheKodeToad commented:
I think three approvals should be enough? @Ryex

Ryex commented:
yeah, we good.
